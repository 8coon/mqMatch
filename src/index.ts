
type MQEvent = 'change';
type MQChangeEventHandler = (matches: string[]) => void;

export interface MQMatchSnapshotRecord {
	raw: string;
	parsed: string;
	matches: boolean;
}

interface QueryRecord {
	value: string;
	query: MediaQueryList;
}

export interface MQMatch {
	/**
	 * Returns the list of current matched queries
	 */
	getCurrentMatches(): string[];

	/**
	 * Returns the current listeners state
	 */
	getCurrentSnapshot(): MQMatchSnapshotRecord[];

	/**
	 * Adds a new media query to the set
	 * If this query matches, fires <change> event
	 * @param mediaQuery
	 */
	register(mediaQuery: string);
	register(key: string, mediaQuery: string);

	/**
	 * Removes given media query from the set
	 * If the removed query matched, fires <change> event
	 * @param key
	 */
	unregister(key: string);

	/**
	 * Subscribes to an event
	 * @param event
	 * @param handler
	 */
	on(event: MQEvent, handler: MQChangeEventHandler);

	/**
	 * Unsubscribes from an event
	 * @param event
	 * @param handler
	 */
	off(event: MQEvent, handler: MQChangeEventHandler);

	/**
	 * Destroys this matcher instance and all its query listeners
	 */
	destroy();
}

class MQMatchImpl implements MQMatch {
	private _queries: Record<string, QueryRecord> = {};
	private _handlers: MQChangeEventHandler[] = [];
	private _lastFiredSnapshot: string;
	private _hackStyleEl: HTMLStyleElement;

	public getCurrentMatches(): string[] {
		return Object
			.keys(this._queries)
			.filter(query => this._queries[query].query.matches)
			.sort();
	}

	public getCurrentSnapshot(): MQMatchSnapshotRecord[] {
		return Object
			.keys(this._queries)
			.map(query => ({
				key: query,
				raw: this._queries[query].value,
				parsed: this._queries[query].query.media,
				matches: this._queries[query].query.matches,
			}));
	}

	public on(event: MQEvent, handler: MQChangeEventHandler) {
		if (event !== 'change') {
			return;
		}

		if (this._handlers.indexOf(handler) === -1) {
			this._handlers.push(handler);
		}
	}

	public off(event: MQEvent, handler: MQChangeEventHandler) {
		if (event !== 'change') {
			return;
		}

		const index = this._handlers.indexOf(handler);

		if (index !== -1) {
			this._handlers.splice(index, 1);
		}
	}

	public register(mediaQuery: string)
	public register(key: string, mediaQuery: string)
	public register(key: string, mediaQuery?: string) {
		if (mediaQuery === undefined) {
			mediaQuery = key;
		}

		if (this._queries[key]) {
			return;
		}

		const query = matchMedia(mediaQuery);

		if (typeof query.addEventListener === 'function') {
			query.addEventListener('change', this._handleMediaQueryChange);
		} else {
			query.addListener(this._handleMediaQueryChange);
		}

		this._queries[key] = {
			value: mediaQuery,
			query,
		};

		if (query.matches) {
			this._handleMediaQueryChange();
		}
	}

	public unregister(key: string) {
		if (!this._queries[key]) {
			return;
		}

		const query = this._queries[key].query;

		if (typeof query.removeEventListener === 'function') {
			query.removeEventListener('change', this._handleMediaQueryChange);
		} else {
			query.removeListener(this._handleMediaQueryChange);
		}

		delete this._queries[key];

		if (query.matches) {
			this._handleMediaQueryChange();
		}
	}

	public destroy() {
		this._handlers.length = 0;

		Object
			.keys(this._queries)
			.forEach(key => this.unregister(key));
	}

	private _handleMediaQueryChange = () => {
		const currentMatches = this.getCurrentMatches();

		const currentMatchesSnapshot = currentMatches.join(',');

		if (this._lastFiredSnapshot !== undefined && this._lastFiredSnapshot === currentMatchesSnapshot) {
			return;
		}

		this._lastFiredSnapshot = currentMatchesSnapshot;

		for (const handler of this._handlers) {
			handler(currentMatches);
		}
	}
}

export function createMediaQueryMatch(): MQMatch {
	return new MQMatchImpl();
}
