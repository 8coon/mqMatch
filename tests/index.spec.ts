import {alter, setup, teardown} from "./mocks/matchMedia.modern";
import {createMediaQueryMatch, MQMatch} from "../src";

describe('Unit', () => {
	let mqMatch: MQMatch;

	beforeAll(() => {
		setup();
		mqMatch = createMediaQueryMatch();
	});

	afterEach(() => {
		alter([]);
	});

	afterAll(() => {
		mqMatch.destroy();
		teardown();
	});

	test('Matches one query', () => {
		mqMatch = createMediaQueryMatch();
		expect(mqMatch.getCurrentMatches()).toEqual([]);

		mqMatch.register('(min-height: 220)');
		alter(['(min-height: 220)']);

		expect(mqMatch.getCurrentMatches()).toEqual(['(min-height: 220)']);

		alter([]);
		expect(mqMatch.getCurrentMatches()).toEqual([]);
	});

	test('triggers events on matched query list change', () => {
		mqMatch = createMediaQueryMatch();
		const handler = jest.fn();

		mqMatch.on('change', handler);

		mqMatch.register('(min-height: 220)');
		alter(['(min-height: 220)']);

		alter([]);
		mqMatch.unregister('(min-height: 220)');

		alter(['(min-height: 220)']);
		mqMatch.register('(min-height: 220)');

		expect(handler.mock.calls).toMatchSnapshot();
	});

	test('multiple queries match at once', () => {
		mqMatch = createMediaQueryMatch();
		const handler = jest.fn();

		mqMatch.on('change', handler);

		mqMatch.register('(min-height: 220)');
		mqMatch.register('(min-height: 440)');

		alter([]);
		alter(['(min-height: 220)']);
		alter(['(min-height: 440)']);
		alter(['(min-height: 220)', '(min-height: 440)']);

		expect(handler.mock.calls).toMatchSnapshot();
	});

	test('on and off works', () => {
		mqMatch = createMediaQueryMatch();
		const handler = jest.fn();

		alter(['(min-height: 220)']);

		mqMatch.on('change', handler);
		mqMatch.on('change', handler);

		mqMatch.register('(min-height: 220)');

		mqMatch.off('change', handler);
		alter([]);

		mqMatch.on('change', handler);
		alter(['(min-height: 220)']);

		mqMatch.off('change', handler);
		mqMatch.off('change', handler);
		alter([]);

		mqMatch.on('change2' as any, handler);
		alter(['(min-height: 220)']);
		mqMatch.off('change2' as any, handler);

		expect(handler.mock.calls).toMatchSnapshot();
	});
});
