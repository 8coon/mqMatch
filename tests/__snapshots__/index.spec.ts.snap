// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit hack for weird matchMedia behavior 1`] = `"@media (min-height: 220px) { .__mqMatchClass__ {} }"`;

exports[`Unit hack for weird matchMedia behavior 2`] = `
"@media (min-height: 440px) { .__mqMatchClass__ {} }
@media (min-height: 220px) { .__mqMatchClass__ {} }"
`;

exports[`Unit hack for weird matchMedia behavior 3`] = `"@media (min-height: 440px) { .__mqMatchClass__ {} }"`;

exports[`Unit multiple queries match at once 1`] = `
Array [
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
  Array [
    Array [
      "(min-height: 440)",
    ],
  ],
  Array [
    Array [
      "(min-height: 220)",
      "(min-height: 440)",
    ],
  ],
]
`;

exports[`Unit on and off works 1`] = `
Array [
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
]
`;

exports[`Unit register and unregister work 1`] = `
Array [
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
  Array [
    Array [],
  ],
]
`;

exports[`Unit returns correct snapshot 1`] = `Array []`;

exports[`Unit returns correct snapshot 2`] = `
Array [
  Object {
    "matches": true,
    "parsed": "(min-height: 220)",
    "raw": "(min-height: 220)",
  },
]
`;

exports[`Unit returns correct snapshot 3`] = `
Array [
  Object {
    "matches": false,
    "parsed": "(min-height: 220)",
    "raw": "(min-height: 220)",
  },
]
`;

exports[`Unit triggers events on matched query list change 1`] = `
Array [
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
  Array [
    Array [],
  ],
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
]
`;

exports[`Unit works with deprecated methods 1`] = `
Array [
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
  Array [
    Array [],
  ],
  Array [
    Array [
      "(min-height: 220)",
    ],
  ],
]
`;
